fml/platform/darwin/scoped_nsobject.h

```diff
@@ -37,29 +37,56 @@ namespace fml {
 // We check for bad uses of scoped_nsobject and NSAutoreleasePool at compile
 // time with a template specialization (see below).
 
+class scoped_nsprotocol_arc_memory_management {
+ public:
+  static id Retain(__unsafe_unretained id object) __attribute((ns_returns_not_retained));
+  static id Autorelease(__unsafe_unretained id object) __attribute((ns_returns_not_retained));
+  static void Release(__unsafe_unretained id object);
+  static id InvalidValue() __attribute((ns_returns_not_retained)) { return nil; }
+};
+
+template <typename NST>
+class scoped_nsprotocol_mrc_memory_management {
+ public:
+  static NST Retain(NST object) { return [object retain]; }
+  static NST Autorelease(NST object) { return [object autorelease]; }
+  static void Release(NST object) { [object release]; }
+  static NST InvalidValue() { return nil; }
+};
+
+#if __has_feature(objc_arc)
```

Comment:
I'm confused by this ifdef; why do you want different ifdef paths on the implementations? That seems like it significantly increases the chances of an accidental ODR violation. The same declarations and implementation of the wrappers should work for both. That was the whole point of the structure of the Chromium change.

In fact, it looks to me like this guarantees ODR violations, since you'll generate two copies of definitions of this template class (for any given `NST`), with different implementations. Am I missing something?

Reply:
ifdef's removed